src/pages/Earn/ClaimModal.tsx

```diff
@@ -0,0 +1,251 @@
+    true
+  );
+  const [shouldStakeEsGmx, setShouldStakeEsGmx] = useLocalStorageSerializeKey(
+    [chainId, "StakeV2-compound-should-stake-es-gmx"],
+    true
+  );
+  const [shouldClaimWeth, setShouldClaimWeth] = useLocalStorageSerializeKey(
+    [chainId, "StakeV2-compound-should-claim-weth"],
+    true
+  );
+  const [shouldConvertWeth, setShouldConvertWeth] = useLocalStorageSerializeKey(
+    [chainId, "StakeV2-compound-should-convert-weth"],
+    true
+  );
+
+  const govTokenAmount = useGovTokenAmount(chainId);
+  const govTokenDelegatesAddress = useGovTokenDelegates(chainId);
+  const isUndelegatedGovToken =
+    chainId === ARBITRUM && govTokenDelegatesAddress === NATIVE_TOKEN_ADDRESS && govTokenAmount && govTokenAmount > 0;
+
+  const gmxAddress = getContract(chainId, "GMX");
+  const stakedGmxTrackerAddress = getContract(chainId, "StakedGmxTracker");
+
+  const [isApproving, setIsApproving] = useState(false);
+
+  const { tokensAllowanceData: gmxAllowanceData } = useTokensAllowanceData(chainId, {
+    spenderAddress: stakedGmxTrackerAddress,
+    tokenAddresses: [gmxAddress],
+  });
+
+  const gmxTokenAllowance = gmxAllowanceData?.[gmxAddress];
+
+  const needApproval =
+    shouldStakeGmx &&
+    totalGmxRewards !== undefined &&
+    ((gmxTokenAllowance !== undefined && totalGmxRewards > gmxTokenAllowance) ||
+      (totalGmxRewards > 0n && gmxTokenAllowance === undefined));
+
+  const isPrimaryEnabled = () => {
+    return !isClaiming && !isApproving && !needApproval && !isUndelegatedGovToken;
+  };
+
+  const getPrimaryText = () => {
+    if (needApproval || isApproving) {
+      return t`Pending GMX approval`;
+    }
+
+    if (isClaiming) {
+      return t`Claiming...`;
+    }
+    return t`Claim`;
+  };
+
+  const onClickPrimary = () => {
+    if (needApproval) {
+      approveTokens({
+        setIsApproving,
+        signer,
+        tokenAddress: gmxAddress,
+        spender: stakedGmxTrackerAddress,
+        chainId,
+      });
+      return;
+    }
+
+    setIsClaiming(true);
+
+    const contract = new ethers.Contract(rewardRouterAddress, RewardRouter.abi, signer);
+    callContract(
+      chainId,
+      contract,
+      "handleRewards",
+      [
+        shouldClaimGmx || shouldStakeGmx,
+        shouldStakeGmx,
+        shouldClaimEsGmx || shouldStakeEsGmx,
+        shouldStakeEsGmx,
+        true,
+        isNativeTokenToClaim ? shouldClaimWeth || shouldConvertWeth : false,
+        isNativeTokenToClaim ? shouldConvertWeth : false,
+      ],
+      {
+        sentMsg: t`Claim submitted!`,
+        failMsg: t`Claim failed.`,
+        successMsg: t`Claim completed!`,
+        successDetailsMsg: !shouldStakeGmx ? gmxUsageOptionsMsg : undefined,
+        setPendingTxns,
+      }
+    )
+      .then(() => {
+        setIsVisible(false);
+        onClaimSuccess?.();
+      })
+      .finally(() => {
+        setIsClaiming(false);
+      });
+  };
+
+  const toggleShouldStakeGmx = (value) => {
```

Comment:
Let's wrap this and following functions into `useCallback`, they depend only on local state

Reply:
done